From 2e56ea86e0584a0ca2830d96591ec6330c9a4d5a Mon Sep 17 00:00:00 2001
From: Jacob Pfeifer <jacob@pfeifer.dev>
Date: Fri, 28 Jul 2023 22:32:15 -0400
Subject: [PATCH] Nudgless Lane Change

Enables automatically changing lanes when the blinker is turned on.
---
 common/params.cc                        |  4 ++++
 selfdrive/controls/lib/desire_helper.py | 13 +++++++++++++
 selfdrive/ui/qt/offroad/settings.cc     |  8 ++++++++
 3 files changed, 25 insertions(+)

diff --git a/common/params.cc b/common/params.cc
index e8ab42c0b..fc8c7310d 100644
--- a/common/params.cc
+++ b/common/params.cc
@@ -206,6 +206,10 @@ std::unordered_map<std::string, uint32_t> keys = {
     {"Version", PERSISTENT},
     {"VisionRadarToggle", PERSISTENT},
     {"WheeledBody", PERSISTENT},
+
+    // PFEIFER - NLC {{
+    {"NudgelessLaneChange", PERSISTENT},
+    // }} PFEIFER - NLC
 };
 
 } // namespace
diff --git a/selfdrive/controls/lib/desire_helper.py b/selfdrive/controls/lib/desire_helper.py
index 4652b41c1..87d7a0bcd 100644
--- a/selfdrive/controls/lib/desire_helper.py
+++ b/selfdrive/controls/lib/desire_helper.py
@@ -1,6 +1,10 @@
 from cereal import log
 from common.conversions import Conversions as CV
 from common.realtime import DT_MDL
+# PFEIFER - NLC {{
+from common.params import Params
+params = Params()
+# }} PFEIFER - NLC
 
 LaneChangeState = log.LateralPlan.LaneChangeState
 LaneChangeDirection = log.LateralPlan.LaneChangeDirection
@@ -67,6 +71,11 @@ class DesireHelper:
         blindspot_detected = ((carstate.leftBlindspot and self.lane_change_direction == LaneChangeDirection.left) or
                               (carstate.rightBlindspot and self.lane_change_direction == LaneChangeDirection.right))
 
+        # PFEIFER - NLC {{
+        if params.get_bool('NudgelessLaneChange'):
+          torque_applied = True
+        # }} PFEIFER - NLC
+
         if not one_blinker or below_lane_change_speed:
           self.lane_change_state = LaneChangeState.off
           self.lane_change_direction = LaneChangeDirection.none
@@ -93,6 +102,10 @@ class DesireHelper:
             self.lane_change_state = LaneChangeState.preLaneChange
           else:
             self.lane_change_state = LaneChangeState.off
+          # PFEIFER - NLC {{
+          if one_blinker and params.get('NudgelessLaneChange'):
+            self.lane_change_state = LaneChangeState.laneChangeFinishing
+          # }} PFEIFER - NLC
 
     if self.lane_change_state in (LaneChangeState.off, LaneChangeState.preLaneChange):
       self.lane_change_timer = 0.0
diff --git a/selfdrive/ui/qt/offroad/settings.cc b/selfdrive/ui/qt/offroad/settings.cc
index 94a673dd7..d1c3fbc9d 100644
--- a/selfdrive/ui/qt/offroad/settings.cc
+++ b/selfdrive/ui/qt/offroad/settings.cc
@@ -69,6 +69,14 @@ TogglesPanel::TogglesPanel(SettingsWindow *parent) : ListWidget(parent) {
       tr("Display speed in km/h instead of mph."),
       "../assets/offroad/icon_metric.png",
     },
+    // PFEIFER - NLC {{
+    {
+      "NudgelessLaneChange",
+      tr("Nudgeless Lane Change"),
+      tr("Enables automatically changing lanes when the blinker is turned on."),
+      "../assets/img_experimental_white.svg",
+    },
+    // }} PFEIFER - NLC
 #ifdef ENABLE_MAPS
     {
       "NavSettingTime24h",
-- 
2.41.0

